Implement loading/saving of Transactionv4

zander opened the pull request:
Allow a transaction of a new version to be loaded using a tagged-format allowing much greater extensibility.

zander committed:
Implement a new transaction format (v4): FlexTrans

This implements the parsing and serialization of v4 transactions
in the format called Compact Message Format.

This fixes the malleability issues in transactions.
This actually shrinks most transactions a couple of procent.
We get linear scaling of sighash operations

Future;
This enables the lightning network to develop.

Next it makes it very easy to fix various problems we know of in Bitcoin;
* far simpler script upgrades.
* fraud proofs.
* pruning signature data in blocks to remove about 75% of the block data.
* reduces bandwidth for light nodes and historical sync

Still work in progress is CSV support.

zander committed:
Fix this test to not randomly generate version numbers.

If it were to create a 4, it would fail.

justusranvier commented on src/consensus/transactionv4.h:
```diff
@@ -0,0 +1,37 @@
+
+enum TxMessageFields {
+    TxEnd = 0,          // BoolTrue
+    TxInPrevHash,       // sha256 (Bytearray)
```
If you added one more field (TxInPrevHeight) to hold the block height at which the previous transaction was mined, that's the only change that would be necessary to make usable fraud proofs.

gmaxwell replied:
> that's the only change that would be necessary to make usable fraud proofs

No its not.

justusranvier replied:
@gmaxwell Did you accidentally post on the wrong thread? It doesn't appear to have any relationship to my comment.

dangquochungne commented on src/consensus/transactionv4.h:
```diff
@@ -0,0 +1,37 @@
+
+enum TxMessageFields {
+    TxEnd = 0,          // BoolTrue
+    TxInPrevHash,       // sha256 (Bytearray)
```
ok.thanks

dangquochungne commented on src/consensus/transactionv4.h:
```diff
@@ -0,0 +1,37 @@
+
+}
+
+#endif
```
hj

greatwolf commented on src/primitives/transaction.h:
```diff
@@ -195,6 +198,59 @@ class CTxOut
+        READWRITE(*const_cast<uint32_t*>(&tx.nLockTime));
+    }
+}
+
```
Why are there so many `const_cast` in this function?

zander replied:
This is historical; the CTransaction uses public const members for whatever stupid reason. Its also the reason there is a CMutableTransaction.

greatwolf commented on src/serialize.h:
```diff
@@ -375,6 +378,160 @@ I ReadVarInt(Stream& is)
+        else if (data.which() == 2)
+            return boost::get<uint64_t>(data);
+        return 0;
+    }
```
Maybe refactor `intData` and `longData` into one template function? eg. `int intData() { return getDataAs<int>(); }`

zander replied:
I'd rather keep the knowledge of what is stored in the variant local in the class. Your suggestion would throw an exaction if a caller didn't exactly use the 2 types of ints supported here.

greatwolf commented on src/serialize.h:
```diff
@@ -375,6 +378,160 @@ I ReadVarInt(Stream& is)
+            if (tmp > 0xFFFFFF)
+                data = (uint64_t) tmp;
+            else
+                data = (int32_t) tmp;
```
What's the reason for preferring C-style cast over C++? Is it just for brevity? Also why the magic number `0xFF FF FF`? Why not `0x7F FF FF FF` or use `numeric_limits<int32_t>::max()`?

zander replied:
i'll fix

zander committed:
Introduce new unit tests to test the new v4 tx.

First move some useful methods to a 'transaction_utils' class.
Introduce testing for
* loading/saving v4 tx.
* Creating a transaction on one node and mining it on another.
* The RPC call createrawtransaction to be able to create a v4 tx
* Signing a v4 tx and being able to validate it properly too
* Reading a hand-created tx (in other software) so we know the
  format hasn't changed

zander commented:
Merged, not sure why it didn't close.
